### ISSUE TITLE
No keyring actions available

### ISSUE THREAD

**Issue by pjmuller** [sylvek/itracing2#4]
Hi Sylvain,

I ordered http://www.aliexpress.com/item/Wireless-Smart-Bluetooth-4-0-Anti-lost-alarm-bluetooth-Tracker-key-finder-Child-Elderly-Pet-Phone/32396376743.html and can pair it in the app but all of the keyring's actions are gray as is the "ring the keyring" action.

Is my token not compatible with your app or am I missing something else?

Greetings,
PJ

**Comment by sylvek** [comment]
Hi Pieterjan Muller :)

All works fine, you just need to activate "brackground link" to allow
actions. :-)

Best regards,
Sylvain

Le lun. 31 août 2015 à 10:33, Pieterjan Muller <notifications@github.com> a
écrit :

**Comment by pjmuller** [comment]
Hi Sylvain,

thanks for the fast answer, true, when I activate the "link background"
toggle, the actions under the "keyring's actions" are not disabled anymore.
However nothing seems to happen, also the "ring the keyring" stays disabled
and I cannot get any battery info.

The strange thing is that the token and the app are paired because when I
turn off my bluetooth, the token starts to beep as it has lost contact...

Any ideas :s :)?

Greetings,
PJ

**Comment by sylvek** [comment]
Okay.. If you see nothing.. It's because your itag device is not found.
When your device will be discovered, the button will available. Try to turn
off / on your itag and your bluetooth...

Le lun. 31 août 2015 18:43, Pieterjan Muller <notifications@github.com> a
écrit :

> Hi Sylvain,
>
> thanks for the fast answer, true, when I activate the "link background"
> toggle, the actions under the "keyring's actions" are not disabled anymore.
> However nothing seems to happen, also the "ring the keyring" stays disabled
> and I cannot get any battery info.
>
> The strange thing is that the token and the app are paired because when I
> turn off my bluetooth, the token starts to beep as it has lost contact...
>
> Any ideas :s :)?
>
> Greetings,
> PJ
>

**Comment by sylvek** [comment]
huumm i read a second time your post. And, something strange occurs.
have you got itracing (1) on your mobile? please try to remove it.
have you paired your itag with your phone via the bluetooth menu? please try to remove it too.

**Comment by pjmuller** [comment]
Hi Sylvain,

thanks for sending me a second mail :p. I did not reply on the mail you
send yesterday because indeed that was not really the problem.
So I tried the things you said

- remove iTracing app
- pairing via bluetooth menu does not work (see screenshot
https://www.dropbox.com/s/bln2c5w04scm5fw/Screenshot_2015-09-01-17-00-23.png?dl=0
)
- iTracing2 does find the token (because if I hit the refresh I always get
an updated rssi info), but the actions don't work (see screenshot
https://www.dropbox.com/s/h7r9a27bizr7kal/Screenshot_2015-09-01-17-00-48.png?dl=0
)

Any other ideas?
The token I have should be compatible with your app right?
http://www.aliexpress.com/item/Wireless-Smart-Bluetooth-4-0-Anti-lost-alarm-bluetooth-Tracker-key-finder-Child-Elderly-Pet-Phone/32396376743.html
looks
exactly the same as Quintic PROXR you mention in the description.

Greetings,
PJ

**Comment by sylvek** [comment]
okay ! i think that i got the answer ... (and first of all, thank you to
analyse the problem instead of adding a comment like "suxx doesn't work")
in fact .. i saw two option.
1st - you select a bad device (maybe you have too much at home and you
select an another device instead of your itag)
2nd - you device is "not" compatible .. but, you could help me to make it :)
please look this wonderfull app (
https://play.google.com/store/apps/details?id=no.nordicsemi.android.mcp&hl=fr
)
connect you to your device and give me a screenshot :)



Le mar. 1 sept. 2015 à 17:04, Pieterjan Muller <notifications@github.com> a
écrit :

> Hi Sylvain,
>
> thanks for sending me a second mail :p. I did not reply on the mail you
> send yesterday because indeed that was not really the problem.
> So I tried the things you said
>
> - remove iTracing app
> - pairing via bluetooth menu does not work (see screenshot
>
> https://www.dropbox.com/s/bln2c5w04scm5fw/Screenshot_2015-09-01-17-00-23.png?dl=0
> )
> - iTracing2 does find the token (because if I hit the refresh I always get
> an updated rssi info), but the actions don't work (see screenshot
>
> https://www.dropbox.com/s/h7r9a27bizr7kal/Screenshot_2015-09-01-17-00-48.png?dl=0
> )
>
> Any other ideas?
> The token I have should be compatible with your app right?
>
> http://www.aliexpress.com/item/Wireless-Smart-Bluetooth-4-0-Anti-lost-alarm-bluetooth-Tracker-key-finder-Child-Elderly-Pet-Phone/32396376743.html
> looks
> exactly the same as Quintic PROXR you mention in the description.
>
> Greetings,
> PJ
>

**Comment by pjmuller** [comment]
Hey Sylvain,

no thank you for helping me :)!
I'm a programmer too (obviously if we communicate through github ;)) so
it's a big pleasure to help you to get this token working too.

I don't think the problem is that I'm pairing with the wrong device. It is
very clear that when I the token on, device MLE-15 appears.
So up to option 2, "not compatible". I went all the way in the app you
suggested and here you have the screenshots

https://www.dropbox.com/sh/iqempbdiouzf8dh/AAD2TwDS4D_OILT492zQRmu7a?dl=0

Greetings,
PJ

**Comment by sylvek** [comment]
Thks! very helpful..
so you are right, second option. But, i don't understand why that doesn't work.
in fact, Bluetooth LE use a protocol based on discoverable "services".
for some reasons of compatibility, some services contain a "public address" like 0x1802 for immediate alert (to ring the keyring), 0x180f to dispatch the battery level .. and "custom address" like 0xffe0 for the keyring button.
but my application uses already those addresses :-/
so after reading my own code (BluetoothLEService.java)… i need more help :-/
i don't know if you have a good skill about android sdk but if you may retrieve the debug trace of itracing2, it would be a very valuable help.

ps.. what kind of smartphone do you have?
re-ps.. i suspect that your mobile phone doesn't inform correctly the software about the connectivity status and the discovering service is not called

**Comment by pjmuller** [comment]
Hi Sylvain,

I'm a ruby developer and do not know anything about the sdk debug traces or
whatever :s.
my device is a Xiaomi redmi note, quite special BUT with iTracing (the
original app) it does work... So then I would not assume it's because of
the phone.

Any other ideas what I could for you in order to test things?

PS: just as a quick question, in iTracing (1), when I turn off bluetooth
the keyring starts beeping, I find this quite annoying. I could turn off
the "alarm" feature on my phone but I would also like to turn it off on the
keyring. Is that possible in iTracing2?
PS2: If the token is not connected to any bluetooth smartphone it shuts off
after 30s, which is quite annoying, do you also have a solution for that in
iTracing2?

Greetings,
PJ

**Comment by sylvek** [comment]
sh** :-/

PS:with iTracing2 you can choose to beep or not your keyring when you're turning off your bluetooth (and the same if you're turning of your keyring your phone may ring or not.
PS2:very strange... i didn't see this feature with my itag :-/ .. my phone auto-reconnect with my keyring when it back.

**Comment by sylvek** [comment]
may i send to you a custom version tomorrow?

**Comment by pjmuller** [comment]
Would be great :)

I'll be your test guinea pig!

**Comment by pjmuller** [comment]
Hi Sylvain, so if I understand correctly, the following situation should
not happen in iTracing 2

I go working in the garden, too far and my token is deconnected, I return
back home and 3 hours later I want to locate my keys BUT by by now the
token is off and I cannot connect to it anymore.
For the record, yes I wish to turn off the "Anti lost" functions because I
do not like my phone to make noises when it's too far from the token."

Long story short, I only want the "alert" functionality from phone ->
token. Even after having being 100's km + a couple days away from my token.
(but now back in a short bluetooth range of course)

Greetings,
PJ

**Comment by sylvek** [comment]
i will do a test with my own itag :)

**Comment by sylvek** [comment]
I have done my test. 
Power on my itag. 
Double click.. It's ringing. 
Power off my itag. 
Power on my itag. 
Wait 20s 
Double click.. It's ringing.

**Comment by pjmuller** [comment]
Hi Sylvain,

sorry for the longer delay, I've had 3 weddings to visit, crazy times.
Thanks for doing the experiment but I'm talking about the other direction.
So trying to let the itag make noise when you click "alert" on your phone.
That does not work when the iTag is off, right? And my iTag turns itself of
after 30s when it was not connected.

So try this if you want

1. pair phone + itag
2. turn bluetooth off on phone
3. Normally the itag starts beeping because it lost contact
4. after 60s long beep to signal that it switches itself off
5. turn on bluetooth on phone
6. Try the alert function, to let the itag beep

With me that does not work, I'll also try out your custom pkg in the train
this evening. I just need to write down all the text that I see in the
bottom of the screen, right?

Thanks for everything :)! Let's make this work!

**Comment by sylvek** [comment]
>So trying to let the itag make noise when you click "alert" on your phone.
>That does not work when the iTag is off, right? And my iTag turns itself of
>after 30s when it was not connected.
wright, i think that the problem comes from your itag :-(

in fact, "turning off" action is not the same that loose connection.. when i'm turning off my phone, the link between my itag/phone is revoked and nothing rings.

>I'll also try out your custom pkg in the train
>this evening. I just need to write down all the text that I see in the
>bottom of the screen, right?

Yes pleaz ! :)

**Comment by paskalip** [comment]
Hi guys. Sylvek did you manage to solve the issue? I got the same? tag device as pjmuller and I'm having the same exact same issues. The app finds the MLE-15, 
![screenshot_2015-10-07-17-34-34](https://cloud.githubusercontent.com/assets/7107440/10344199/d23f56e8-6d19-11e5-95d9-93169ef80e2f.png)
![screenshot_2015-10-07-17-34-23](https://cloud.githubusercontent.com/assets/7107440/10344200/d2428cc8-6d19-11e5-8b34-b41fe116501d.png)

**Comment by sylvek** [comment]
Hi guys, thanks for your report. I'll check a second time tomorrow.

**Comment by paskalip** [comment]
Have you had any time to troubleshoot?

**Comment by sylvek** [comment]
Not yet sorry.. May I send you a specific version with visible toasts to
describe what 's going on?

**Comment by paskalip** [comment]
Sure, whatever I can do to help

**Comment by sylvek** [comment]
thks! i'm going to send you a mail with an attached apk.
this apk contains itracing2 with some traces visible on your screen..
please send me what you read.

**Comment by sylvek** [comment]
please send me a mail at smaucourt@gmail.com and i'ill send you back this apk.

**Comment by paskalip** [comment]
The new apk goes until "connected() - connecting GATT" and stalls there with the update circle going round and round at the top. Now I've got no rssi info as well. I reinstalled itracing just to make sure the keyring is working and it does connect and function properly.

**Comment by sylvek** [comment]
hummm… damned ok thks :-/ sorry to loose your time.
i'm going to try a new way.

**Comment by paskalip** [comment]
Are you serious? I'm not loosing any time. You are spending your time building an app that I want to use. I'm more than happy to assist!

**Comment by sylvek** [comment]
thks man!
i'm sending you a new version .. this time, that should write a file "itracing2.txt" on your sdcard... 
play with itracing2 and send me it please :) :)

**Comment by paskalip** [comment]
Connected, enabled-disabled "link background", changed keyring's actions,
pressed every button, pressed keytag's button, did few refreshes (rssi
works now), closed and reopened app. Text file is mocking me...

On Thu, Oct 15, 2015 at 4:44 PM, Sylvek <notifications@github.com> wrote:

> thks man!
> i'm sending you a new version .. this time, that should write a file
> "itracing2.txt" on your sdcard...
> play with itracing2 and send me it please :) :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sylvek/itracing2/issues/4#issuecomment-148428304>.
>

onServicesDiscovered()

**Comment by sylvek** [comment]
Huuumm ok I will check tonight.. Sorry I do this with my dirty fingers....
Grr

**Comment by hosek** [comment]
Hi all, 
I have the same MLE-15 keyring, just forked the repo and debug it a little (nice app btw!)
Problem was with  BluetoothGattDescriptor descriptor = characteristic.getDescriptors().get(0); as this keyring has 0 descriptors, this leads to crash and endless "refreshing"
After handling it, hidding "refreshing" indicator after any service discovered, it partialy works (no button actions) 

 @pjmuller : I just tried your case - shield device in metal can, it starts beeping, wait for silence, open can and press "ring keyring" in mobile app, device starts the alarm... no turning itselfs off.

**Comment by sylvek** [comment]
Awesome!!!

**Comment by mofosyne** [comment]
Hmmmm.... hosek, if it works for you, I can try and test if it works for me as well. Maybe lets merge it to this repo and see what we get.

**Comment by sylvek** [comment]
:+1: i saw that you migrated sources to gradle thanks a lot!

**Comment by hosek** [comment]
I will make a pull request once I will be sure that all changes are necessary and working - probably over this weekend.
Strange that the batery service has right id, even the nordicsemi app could read it. 
And one very bad feature of this keyring - when it lost signal, it starts to beep for one minute, after that it probably waits for reconection (no automatic reconnection) and then after next cca 3 minutes without signal, it turns itselfs off :/ So its unusable - you have only 3 minutes to find the tag! I hope that this could be changed by some settings - has anybody tried it with the original app?
Thanks

**Comment by mofosyne** [comment]
Well if it means anything. I'm trying to talk to these manufactures to work
out what the chip is in it, sooo that maybe we can grab the firmware, or
maybe even the source code.

Thought... I'm more thinking of the possibility of a cheapass BLE beacon.

**Comment by sylvek** [comment]
>you have only 3 minutes to find the tag! I hope that this could be changed by some settings - has anybody tried it with the original app?

:-o .... i'ill try with mine (i never lost my iTag ;)

**Comment by sylvek** [comment]
>Well if it means anything. I'm trying to talk to these manufactures to work
out what the chip is in it, sooo that maybe we can grab the firmware, or
maybe even the source code.

that should be great :+1:

**Comment by sylvek** [comment]
>I will make a pull request once I will be sure that all changes are necessary and working - probably over this weekend.

good, that let me a few time to fix this issue

>Strange that the batery service has right id, even the nordicsemi app could read it, but actual code cannot read it. 

i think that nordic apps use a different way to discover services :-(

**Comment by hosek** [comment]
Ok, battery info is working:
```java
 batteryCharacteristic = service.getCharacteristics().get(0);
 gatt.readCharacteristic(batteryCharacteristic);
```
Is the app fully working with your tag? As I dont want to break the compatibility with your tag, I have only the MLE-15 one.

**Comment by sylvek** [comment]
Yes. That's working but...  I think that you must subscribe to value changes for your device.. It's why you don't have value.

**Comment by sylvek** [comment]
My device contains a default value and if I want to retrieve the battery level in real-time I should subscribe it. (I think that your device doesn't have a default value)

**Comment by hosek** [comment]
If I understand correctly, only button service needs to be subscribed for my device, get the clicks. Battery has "normal" descriptor - sorry for bad naming, I just started with BLE.
Could you please test this build if its working correctly with your iTag?
https://drive.google.com/file/d/0B0xPH953MoR6d08xb0NfZmdUQlE/view?usp=sharing

**Comment by sylvek** [comment]
humm strange .. no.. battery level is empty :-/

Le mar. 27 oct. 2015 à 21:43, Roman Hosek <notifications@github.com> a
écrit :

**Comment by sylvek** [comment]
in fact some characteritic needs to be subscribed to retrieve data.
like this code:

if (service.getUuid().toString().startsWith(BUTTON_PREFIX)) { final
BluetoothGattCharacteristic characteristic = service.getCharacteristics().
get(0); final BluetoothGattDescriptor descriptor = characteristic.
getDescriptors().get(0); gatt.setCharacteristicNotification(characteristic,
true); descriptor.setValue(BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
); gatt.writeDescriptor(descriptor); }

Le mar. 27 oct. 2015 à 21:55, Sylvain Maucourt <smaucourt@gmail.com> a
écrit :

**Comment by sylvek** [comment]
all line are needed and an event will occur on "onCharacteristicChanged"

Le mar. 27 oct. 2015 à 21:57, Sylvain Maucourt <smaucourt@gmail.com> a
écrit :

**Comment by hosek** [comment]
And other functions are ok?

**Comment by sylvek** [comment]
this release should fix this issue…
https://drive.google.com/file/d/0B2WJ90QJfme5SVMzZ3J6aDNqX2c/view?usp=sharing

**Comment by Forever-Young** [comment]
So, is this scenario possible after all: https://github.com/sylvek/itracing2/issues/4#issuecomment-137473586 ?

I've got MLE-15 tag too, original app (looks like https://play.google.com/store/apps/details?id=com.fb.antiloss but with some changes, downloaded from QR-link specified in docs) doesn't see it at all (neither from store, nor from their link).

My intention is just to start alarm at itag when I need it by pressing a button in app.

**Comment by sylvek** [comment]
hi @Forever-Young , with a mle-15 it seems impossible, sorry :(

**Comment by biotgsj** [comment]
Hi,
using nRF Master Control Panel (or any other iTag compatible app)
service 0x180f to dispatch the battery level, but always shows the same value = 99%
even when battery drains ... any ideas? Or alternative beacons that do work on bat level? 
(or temperature)

**Comment by sylvek** [comment]
Again.. Measuring the battery voltage is not so easy.. It depends on ambiant temperature, battery quality, etc. You have around 3 month of battery so the value doesn't change quickly.

**Comment by biotgsj** [comment]
I intentionnaly lowered the battery voltage, no change! stuck at 99%  (another key is stuck at 78%)

**Comment by sylvek** [comment]
my itag oscillates from 30 to 40% …

**Comment by biotgsj** [comment]
at least you have some change ... bizarre no?  could you change the battery with a new one?
Anyone knows how to "activate" this and eventually also the temperature readout, as some tags offer this service ... interested to know which ones do !
